internal/compiler/program.go

```diff
@@ -174,6 +171,75 @@ func NewProgram(options ProgramOptions) *Program {
 	return p
 }
 
+func NewProgramWithChangedFile(old *Program, changedFilePath tspath.Path) *Program {
+	oldFile := old.filesByPath[changedFilePath]
+	newFile := old.host.GetSourceFile(oldFile.FileName(), changedFilePath, oldFile.LanguageVersion)
+	if !canReplaceFileInProgram(oldFile, newFile) {
+		return NewProgram(old.programOptions)
+	}
+	p := &Program{
+		host:                         old.host,
+		programOptions:               old.programOptions,
+		compilerOptions:              old.compilerOptions,
+		configFileName:               old.configFileName,
+		nodeModules:                  old.nodeModules,
+		currentDirectory:             old.currentDirectory,
+		configFileParsingDiagnostics: old.configFileParsingDiagnostics,
+		resolver:                     old.resolver,
+		comparePathsOptions:          old.comparePathsOptions,
+		processedFiles:               old.processedFiles,
+		filesByPath:                  old.filesByPath,
+		currentNodeModulesDepth:      old.currentNodeModulesDepth,
+		usesUriStyleNodeCoreModules:  old.usesUriStyleNodeCoreModules,
+		unsupportedExtensions:        old.unsupportedExtensions,
+	}
+	p.initCheckerPool()
+	index := core.FindIndex(p.files, func(file *ast.SourceFile) bool { return file.Path() == newFile.Path() })
+	p.files = slices.Clone(p.files)
+	p.files[index] = newFile
+	p.filesByPath = maps.Clone(p.filesByPath)
+	p.filesByPath[newFile.Path()] = newFile
+	return p
+}
+
+func (p *Program) initCheckerPool() {
+	if p.programOptions.CreateCheckerPool != nil {
+		p.checkerPool = p.programOptions.CreateCheckerPool(p)
+	} else {
+		p.checkerPool = newCheckerPool(core.IfElse(p.singleThreaded(), 1, 4), p)
+	}
+}
+
+func canReplaceFileInProgram(file1 *ast.SourceFile, file2 *ast.SourceFile) bool {
+	return file1.FileName() == file2.FileName() &&
+		file1.Path() == file2.Path() &&
+		file1.LanguageVersion == file2.LanguageVersion &&
+		file1.LanguageVariant == file2.LanguageVariant &&
+		file1.ScriptKind == file2.ScriptKind &&
+		file1.IsDeclarationFile == file2.IsDeclarationFile &&
+		file1.HasNoDefaultLib == file2.HasNoDefaultLib &&
+		file1.UsesUriStyleNodeCoreModules == file2.UsesUriStyleNodeCoreModules &&
+		slices.EqualFunc(file1.Imports, file2.Imports, compareImports) &&
+		slices.EqualFunc(file1.ModuleAugmentations, file2.ModuleAugmentations, compareModuleAugmentations) &&
+		slices.Equal(file1.AmbientModuleNames, file2.AmbientModuleNames) &&
+		slices.EqualFunc(file1.ReferencedFiles, file2.ReferencedFiles, compareFileReferences) &&
+		slices.EqualFunc(file1.TypeReferenceDirectives, file2.TypeReferenceDirectives, compareFileReferences) &&
+		slices.EqualFunc(file1.LibReferenceDirectives, file2.LibReferenceDirectives, compareFileReferences) &&
+		file1.CheckJsDirective == file2.CheckJsDirective
```

Comment:
These are pointers; probably needs to be a more complicated function which checks pointer equality, nilness, and the underlying values?

Reply:
It sure does.